### PR TITLE
[docs] Add MessageListener example in addition to endless while loop

### PR DESCRIPTION
**Pull request by newur**
### Motivation

Using a MessageListener is simpler than using a endless while loop. The loop might work for super basic hello-world examples, but would trap any other app with its endless nature. Also the loop might motivate users to move the while loop into an own, custom build thread, which is also unnecessary complicated.

### Modifications

Replace the `while` loop with a `MessageListener`.

### CODE REVIEW

File: site2/docs/client-libraries-java.md [hunk 1]
```diff
@@ -243,16 +243,14 @@ Once you've instantiated a {@inject: javadoc:PulsarClient:/client/org/apache/pul
...
         .subscriptionName("my-subscription")
+        .messageListener(myMessageListener)
         .subscribe();
 ```
 
-The `subscribe` method will auto subscribe the consumer to the specified topic and subscription. One way to make the consumer listen on the topic is to set up a `while` loop. In this example loop, the consumer listens for messages, prints the contents of any received message, and then [acknowledges](reference-terminology.md#acknowledgment-ack) that the message has been processed. If the processing logic fails, you can use [negative acknowledgement](reference-terminology.md#acknowledgment-ack) to redeliver the message later.
+The `subscribe` method will auto subscribe the consumer to the specified topic and subscription. One way to make the consumer listen on the topic is to add a `MessageListener`. In this example, the consumer listens for messages, prints the contents of any received message, and then [acknowledges](reference-terminology.md#acknowledgment-ack) that the message has been processed. If the processing logic fails, you can use [negative acknowledgement](reference-terminology.md#acknowledgment-ack) to redeliver the message later.
```
Comment (Anonymitaet) [inline site2/docs/client-libraries-java.md:250]: @BewareMyPower could you please help review? Thanks

**Comment by Anonymitaet** [comment]
@newur thanks for your contribution. Does this affect only master or other versioned docs? 
If latter, could you please help update all affected versions? Thanks

**Comment by newur** [comment]
@Anonymitaet the versioned docs are the content in `site2/website/versioned_docs`?

Sure I would update them when the change gets approval. How far back do you normally port the docu back? There are quite some :D

**Comment by Anonymitaet** [comment]
Yes, all versioned docs are stored in [pulsar/site2/website/versioned_docs/](https://github.com/apache/pulsar/tree/master/site2/website/versioned_docs). You can "batch and replace" all occurrences with just a few clicks.

**Comment by BewareMyPower** [comment]
Using `receive`/`receiveAsync` and using `MessageListener` are two different ways for consumer, I think it's better to add a section for the example code that uses `MessageListener`, not just replace the `receive` way with `MessageListener`.

**Comment by newur** [comment]
@BewareMyPower what characterizes the difference? What value does it add to keep the `while` example?

**Comment by newur** [comment]
> You can "batch and replace" all occurrences with just a few clicks.

Can you please tell me more about this? :)

**Comment by BewareMyPower** [comment]
@newur It's a synchronous vs. asynchronous compare. The synchronous way that uses `receive` method is easier to use for newbies. I've seen many newbies that tried to use the asynchronous way that sets the listener while they didn't realize that they should do something in the main thread.

```java
            Consumer<String> consumer = client.newConsumer(Schema.STRING)
                    .subscriptionName("my-sub")
                    .topic("my-topic")
                    .messageListener(new MessageListener<String>() {
                        @Override
                        public void received(Consumer<String> consumer, Message<String> msg) {
                            System.out.println("Receive " + msg.getValue() + " from " + msg.getMessageId());
                        }
                    })
                    .subscribe();
            // TODO: do something else
            // If you don't do anything to block here, the consumer will close immediately.
            consumer.close();
```

Some newbies might think when they set the listener, the consumer will block until some state. You may think they're stupid, but the **stupid** guys exist.

Asynchronous programming is more complicated than synchronous programming. Given a case that you need to collect 10 messages to a list and then close the consumer. Using `receive` is easy.

Using listener is somehow complicated:

```java
            final CountDownLatch latch = new CountDownLatch(10);
            final List<String> messages = Collections.synchronizedList(new ArrayList<>()); // [1]
            Consumer<String> consumer = client.newConsumer(Schema.STRING)
                    .subscriptionName("my-sub")
                    .topic("my-topic")
                    .messageListener(new MessageListener<String>() {
                        @Override
                        public void received(Consumer<String> consumer, Message<String> msg) {
                            if (messages.size() < 10) { // [2]
                                messages.add(msg.getValue());
                                consumer.acknowledgeAsync(msg); // just to avoid catching the exception
                            }
                            latch.countDown();
                        }
                    })
                    .subscribe();
            latch.await(); // [3]
            messages.forEach(System.out::println); // Users may do something more complicated here
            consumer.close();
```

1. You should use a thread-safe container because you'll access these messages in both main thread and listener thread.
2. Since you cannot control whether to stop calling the listener, you need to check the count of received messages in case the consumer received more than 10 messages internally.
3. You need an extra component like `CountDownLatch`  for the synchronization between the main thread and the listener thread.

While using `receive` method is much easier:

```java
            final List<String> messages = new ArrayList<>();
            Consumer<String> consumer = client.newConsumer(Schema.STRING)
                    .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
                    .subscriptionName("my-sub")
                    .topic("my-topic")
                    .subscribe();
            for (int i = 0; i < 10; i++) {
                // NOTE: If you're using receive() with timeout, you must check null for timeout
                messages.add(consumer.receive().getValue());
            }
            messages.forEach(System.out::println);
            consumer.close();
```

**Comment by BewareMyPower** [comment]
> Also the loop might motivate users to move the while loop into an own, custom build thread, which is also unnecessary complicated.

Yeah, you might think my example before is just to make it complicated. But using listeners also might motivate users to wrap all complicated logic into the listener. Just like what Pulsar does, for each method, Pulsar provides both the asynchronous `xxxAsync()` method and the synchronous `xxx()` method. Users can choose the suitable one for their own scenarios. We can't assume either one is always better. The point also goes for `receive` vs. `messageListener` compare.

IMO, adding the missed documents for how to use listeners for consumers is good. Removing the `receive` example is bad except the `receive` method was marked as deprecated in future.

**Comment by newur** [comment]
@BewareMyPower in my opinion this is a very different and more advanced use case that supports your claim. I read the current doc in the spirit of 'Look, here is a basic example how you can fetch messages one by one'.

But the example has the major drawback that it traps the app (main thread) so it can never do anything else than receiving messages. This is a limitation that goes far beyond what I would call reasonable. The user will most likely want its app to be not trapped (like I would like my SpringBoot/Quarkus/whatever app to finish the normal startup).

So the listener approach is similar complex to read/write to current one, without the trap thingy.

```
            final List<String> messages = new ArrayList<>();
            Consumer<String> consumer = client.newConsumer(Schema.STRING)
                    .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
                    .subscriptionName("my-sub")
                    .topic("my-topic")
                    .subscribe();
            for (int i = 0; i < 10; i++) {
                // NOTE: If you're using receive() with timeout, you must check null for timeout
                messages.add(consumer.receive().getValue());
            }
            messages.forEach(System.out::println);
            consumer.close();
```
If I read that right this `receive` call is blocking, so either I get a timeout or the app is stuck till the message arrives. Again I would never want/expect that behavior my app. Plus, this use case is on a totally different level of complexity, compared to what is currently done.

Maybe you can outline a bit more when the current `while` loop would be desirable. I totally fail to see any use case where I would prefer to have my main thread blocked while consuming messages one by one.

**Comment by BewareMyPower** [comment]
IMO, in short, the `MessageListener` should be the advanced API that's used for efficiency if users prefer asynchronous API. But the `receive` method should be the easiest API that's suitable for basic example.

> If I read that right this receive call is blocking, so either I get a timeout or the app is stuck till the message arrives. 

Not completely right. If there're already messages that are cached in the consumer, the `receive` method will return immediately. It looks like you strongly dislike the synchronous `receive` API that may block the current thread. However, let's look at Kafka's API design: https://kafka.apache.org/28/javadoc/org/apache/kafka/clients/consumer/Consumer.html#poll(java.time.Duration)

It provides the synchronous API (`poll()` method) directly. What's different from Pulsar's `receive()` API is that Kafka's `poll()` returns a container of messages while Pulsar's `receive()` returns a single message. However, Pulsar's API design just avoids collecting the messages into a container. If there're some messages that are cached, the client will only fetch a single message from the cache and pass it to user side. Therefore, in essential they're similar.

- Kafka: fetch N messages (block main thread for some time), collect to a list and return the list to the user.
- Pulsar: fetch N messages (block main thread for some time), users peek messages one by one.

> I totally fail to see any use case where I would prefer to have my main thread blocked while consuming messages one by one.

Maybe it's right to you. But lots of users uses Kafka's synchronous `poll()` API at the same time. You can also see the example in Kafka's official website: https://kafka.apache.org/28/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html

```java
     while (true) { // infinite loop
         ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(100)); // block main thread for some time
         for (ConsumerRecord<String, String> record : records) // peek messages one by one
             System.out.printf("offset = %d, key = %s, value = %s%n", record.offset(), record.key(), record.value());
     }
```

Maybe some Kafka client's wrapper like SprintBoot will provide a listener like API that has some overhead. In this case, Pulsar provides the `MessageListener` that can avoid users from wrapping a listener in another thread again.

**Comment by newur** [comment]
Looks like we agree to disagree on this topic.

For me 'Kafka is doing the same' is not a use case and I hope Pulsar strive to provide a better user experience than Kafka. I would even assume (of course I can not prove that point) that Kafka is in kind of an advantage here, because all major frameworks provide an integration that abstracts this problem fully away (with annotations and the like). As long as pulsar does not have that, we need to develop other strategies, like better docu. :)

Anyway, I would like to follow the suggestion you made and just have both cases in the docu. Not sure if I have time today, so probably later this week I can provide another proposal.

> It looks like you strongly dislike the synchronous receive API that may block the current thread.

I have yet to meet a developer that likes his main thread to be block. :D

**Comment by newur** [comment]
> If we really want to show some best practices the community is using when building application with Pulsar we can add a new section with real world code snippet of how Pulsar API is used in production to produce/consume message. @MarvinCai 

That sounds awesome. I would assume their is a high demand for such a section. What would you consider to be useful measures to create and populate such a section? 

(I cannot really imagine the kind of circumstances where a non-enterprise person/group even dares to deal with pulsar/kafka and such stuff. Not saying that this does not happen. :D)

**Comment by Anonymitaet** [comment]
@BewareMyPower @MarvinCai could you please take a look at the new changes?

**Review by BewareMyPower — APPROVED** [review]
LGTM

**Comment by Anonymitaet** [comment]
@newur thanks for your contribution. Does this affect only master or other versioned docs? 
If latter, could you please help update all affected versions? Thanks

**Comment by BewareMyPower** [comment]
@newur FYI, https://github.com/apache/pulsar/pull/10988 is an example. The documents of older pulsar versions are under `site2/website/versioned_docs` directory.

**Comment by newur** [comment]
I updated the versioned docs.

**Comment by Anonymitaet** [comment]
> I updated the versioned docs. @BewareMyPower @Anonymitaet

@newur thanks, 2.8.1 doc does not need to be updated?

**Comment by newur** [comment]
This branch is behind master, so it does not know 2.8.1. I can do a rebase or create a new PR, whatever you would prefer @Anonymitaet

**Comment by Anonymitaet** [comment]
@newur thanks for your contribution!

**Pull request merged**